### DISCUSSION TITLE
headers in the body of the HTTP message?

### DISCUSSION THREAD

**Discussion by NRollo** [cesanta/mongoose#1576] [Ask questions, get answers] [answered]
Looking in the HTTP message body section from and upload form it seems that there are a few headers present:
```
------WebKitFormBoundaryanTJzIN6drDwbwaj
Content-Disposition: form-data; name="update"; filename="file.ini"
Content-Type: application/octet-stream

# Hashtag = comment
# [????] = section mark
.
.
.
```
I would have thought that they would have been part of the 'Header' section and not the body section of the HTTP message??

**Comment by cpq** [comment, marked as answer]
That's how the body of multipart-encoded message looks like!
See https://mongoose.ws/documentation/#struct-mg_http_part

**Reply by NRollo** [reply]
@cpq Thanks for the prompt answer!
That make sense, - then there is something I don't understand with the form upload. If I upload a "small" file it will be handled by Mongoose as an MG_EV_HTTP_MSG event but when the file is "big" it uses the MG_EV_HTTP_CHUNK  event, it's the same form being used. So, should I use the mg_http_next_multipart(), as shown in the "form-upload" example, in the MG_EV_HTTP_CHUNK event function as well?

Does the "old" mg_http_upload() still work as in release 7.4? I can't find it in the documentation nor in the examples. That was what triggered me to investigate the "form-upload" and the "file-upload" examples.

**Reply by scaprile** [reply]
The **form-upload** example uses _multipart-form-data_ encoding, and assumes the whole POST transfer will fit in memory
```html
      <div id="wrapper">
        <form action="/upload" method="post" enctype="multipart/form-data">
```
You can get all parts you submit using `mg_http_next_multipart()` as the example shows

The **file-upload** example does not use _multipart-form-data_ encoding, it uses a small JavaScript code that uses the `fetch()` browser API to send the file. As you see in what you pasted, it is using _form-data_. It assumes the file won't fit in memory, hence a multipart encoding won't fit either and you wouldn't be able to jump from part to part.
The example also sends the file name in the query string:
```html
fetch('/upload?name=' + encodeURIComponent(f.name), {
```
and shows how you can get it, as explained [here](https://github.com/cesanta/mongoose/discussions/1574#discussioncomment-2873354), though you can parse the header inside the POST content as you already did (I don't know if something could be browser dependent inside that).

**Reply by cpq** [reply]
Thanks @scaprile !

@NRollo , there currently 3 file upload methods.
1. Using multipart upload. The whole file is sent as a  one blob as a multipart POST request. That's the form-upload example
2. Using plain POST method. The whole file is sent as a one blob as a binary POST. A server side does not wait until the whole file is buffered, but processes the file as chunks arrive. That's the file-upload example
3. Using a sequence of small POST requests. A client in this case must take the file, split it by chunks, and send to a server using binary POST and name/offset in the query string. That is "non-standard". And that is what mg_http_upload() handles. Since it is not standard and requires some special handling on the client side, it is removed from the examples. It is not yet decided whether to deprecate it or not.

**Reply by NRollo** [reply]
@cpq Thanks, that makes it a whole lot clearer to me.
I will continue my effort in updating my application in order not to use the mg_http_upload() API.
